### PR TITLE
Introduce Elasticsearch readiness probe

### PR DESCRIPTION
**Pull request by lukas-vlcek**
Simple readiness probe script for Elasticsearch.

Right now the script return success if Elasticsearch can accept traffic on HTTPs port.
I think later we shall consider making the readiness probe pass only if ES has been correctly initialized - this means only after we can verify ES contains expected index templates, SearchGuard has been seeded ... etc.

### CODE REVIEW

**Comment by lukas-vlcek** [comment]
@jcantrill @ewolinetz please add "Do not merge" label for now.

File: elasticsearch/probe/readiness.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# TODO: we need to re-use code that is currently found in ./run.sh
+ES_REST_BASEURL=https://localhost:9200
+secret_dir=/etc/elasticsearch/secret
+max_time=300
```
Comment (richm) [inline elasticsearch/probe/readiness.sh]: Do we need a timeout here, or will that be handled at the higher level e.g. the replication controller will run this script every N seconds, and will report a hard failure if it did not receive a successful reply after M seconds?

**Comment by lukas-vlcek** [comment]
Just for the record how probes are implemented: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/prober/worker.go
(though we should check Kubernetes version used in Openshift)

File: elasticsearch/probe/readiness.sh (hunk 1 above)
Comment (lukas-vlcek) [inline elasticsearch/probe/readiness.sh]: I think the `readinessProbe` can be [configured with `timeoutSeconds`](https://docs.openshift.org/latest/dev_guide/application_health.html):

> How long to wait for the probe to finish (default: 1). If this time is exceeded, OpenShift Origin considers the probe to have failed.

But may be we should still try to terminate the curl command if it runs to long regardless Kubernetes timeouting the probe test? May be we should just make sure that the curl timeout is not set shorter than the probe `timeoutSeconds`?

Comment (richm) [inline elasticsearch/probe/readiness.sh]: I think the timeout in this context should be `1`.  Perhaps something like this:
```
max_time=${max_time:-1}
```
and let k8s just retry periodically until it hits its max timeout or max failures.

**Comment by lukas-vlcek** [comment]
Simple readiness probe for Elasticsearch. @richm @jcantrill PTAL

Right now the script return success if Elasticsearch can accept trafic on HTTPs port.
I think later we shall consider making the readiness probe pass only if ES has been correctly initialized - this means only after we can verify ES contains expected index templates, SearchGuard has been seeded ... etc.

**Comment by richm** [comment]
lgtm

**Comment by jcantrill** [comment]
[merge]

**Comment by jcantrill** [comment]
@lukas-vlcek I might have added merge tag early.  How does this probe get into the right place since I don't see a change in how we build the image.  https://github.com/openshift/openshift-ansible/pull/3733 declares its use but it wont get into the appropriate location.

**Comment by ewolinetz** [comment]
@jcantrill @lukas-vlcek  i aborted the merge test

**Comment by lukas-vlcek** [comment]
@ewolinetz @jcantrill You are right! I added this file to Dockerfile

File: elasticsearch/Dockerfile [hunk 2]
```diff
@@ -37,6 +37,7 @@ RUN rpm --import https://packages.elastic.co/GPG-KEY-elasticsearch && \
 
 ADD sgconfig/ ${HOME}/sgconfig/
 ADD index_templates/ /usr/share/elasticsearch/index_templates/
+ADD probe/readiness.sh /usr/share/elasticsearch/probe/readiness.sh
```
Comment (ewolinetz) [inline elasticsearch/Dockerfile]: i'm not 100% but you may need to create the `probe` sub-directory first... travis should verify though

**Comment by ewolinetz** [comment]
@lukas-vlcek i think you may need to rebase

**Comment by lukas-vlcek** [comment]
@ewolinetz ok, did one more improvement (we are adding whole /probe folder instead of single script in that folder) and rebased.

**Comment by ewolinetz** [comment]
don't listen to rosie, i can do that after travis completes

**Comment by ewolinetz** [comment]
[test]

**Comment by ewolinetz** [comment]
https://ci.openshift.redhat.com/jenkins/job/test-origin-aggregated-logging/1237/

**Comment by ewolinetz** [comment]
[merge]

**Comment by lukas-vlcek** [comment]
@ewolinetz so [that](https://ci.openshift.redhat.com/jenkins/job/test-origin-aggregated-logging/1237/) was false alarm?

**Comment by ewolinetz** [comment]
@lukas-vlcek the one i kicked off manually failed, im curious if rosie will pick it back up now; if not i'll kick off another manual job

**Comment by ewolinetz** [comment]
@lukas-vlcek this merge test failed in the same place the manual one did... it was unable to detect ES started from the pod's logs.

I believe https://github.com/openshift/origin-aggregated-logging/pull/367 will fix it

**Comment by richm** [comment]
[merge]

**Comment by richm** [comment]
[merge]

**Pull request merged**